Add resumé blog entry with mods to accommodate (#40)
briandominick/codewriting: _posts/2019-01-08-data-driven-static-resume.adoc

= Resumé-as-Code: Small Data Sourcing for Fun & Profit
:page-layout: post
:page-permalink: blog-small-data-driven-resume
:page-date: 2019-01-09 17:22
:page-comments: true
:page-external-url:
:page-categories: [small-data]
:page-tags: [small data,Liquid,AsciiDoc,PDF,single sourcing,DRY]
:page-author: Brian Dominick
:toc: macro

> What better way to show off what I do as a software documentation platform developer (“DocOps engineer”) than by generating my own resumé with the technology stack I use professionally?

Reviewer: ```suggestion
What better way to show off what I do as a documentation platform developer (“DocOps engineer”) than by generating my own resumé with the technology stack I use professionally?
```

Several prospective clients have remarked on the demonstrative novelty of this approach.
If you also earn a crust from setting up documentation environments for engineering teams and the like, consider building your professional documentation from a single source to show off your skills.

With its singular source and dual link:https://www.codewriting.org/brian-dominick-resume[HTML] and link:https://www.codewriting.org/assets/files/brian-dominick-resume.pdf[PDF] output, my resumé is a simple model implementation of the AJYL tech stack, consisting of *AsciiDoc*, *Jekyll*/*JAMstack*, *YAML*, and *Liquid*.

In this post, I will walk you through how I set it all up, explaining how each component language and utility come into play.

toc::[]

== The Sources

> These two output versions (link:https://www.codewriting.org/brian-dominick-resume[HTML] | link:https://www.codewriting.org/assets/files/brian-dominick-resume.pdf[PDF]) are sourced from the exact same files, so they are always in sync;

Reviewer: ```suggestion
These two output versions (link:https://www.codewriting.org/brian-dominick-resume[HTML] | link:https://www.codewriting.org/assets/files/brian-dominick-resume.pdf[PDF]) are sourced from the exact same files, so they are always in sync.
```

> I do not have to remember to update two documents whenever I make a change.

Reviewer: Redundant with what comes before; I would actually remove this sentence entirely.

> Most of the specific content items are stored as _small data_ objects arranged in a link:https://idratherbewriting.com/documentation-theme-jekyll/mydoc_yaml_tutorial.html[YAML] file, similarly to fields in a database.

Reviewer: ```suggestion
Most of the specific content items are stored as _small data_ objects arranged in a link:https://idratherbewriting.com/documentation-theme-jekyll/mydoc_yaml_tutorial.html[YAML] file, similar to fields in a database.
```

Author: Awesome catch. The _objects are similar to their counterparts in databases_ -- that is how most people will read this, no doubt. In my mind, it said they're _arranged similarly to their counterparts_.

[source,yaml]
----
include::data/dominick.yml[tags="example-snippet"]
----

If you inspect the link:https://github.com/briandominick/codewriting/blob/master/data/dominick.yml[whole file], you will see it contains several clusters of data, each analogous to a table in a database.

[TIP]
One advantage of maintaining small data in flat files is that you can build an entire, perfectly portable datasource with a text editor in mere minutes.

In this case, parameter values contain link:https://asciidoctor.org/docs/what-is-asciidoc/[AsciiDoc] markup.
Notice on the first line beginning with a `destination:` key, the value text contains AsciiDoc hyperlink markup.

This data gets parsed with a link:https://github.com/Shopify/liquid/wiki[Liquid] template.
This particular Liquid template (link:https://raw.githubusercontent.com/briandominick/codewriting/master/_templates/liquid/resume.asciidoc[see complete source]) weaves those AsciiDoc-formatted strings from the data file with more AsciiDoc markup.

.From resume.asciidoc -- Template mixes Liquid with AsciiDoc, generates new AsciiDoc source
[source]
----
include::assets/includes/resume.asciidoc[tags="employment"]
----

> Because Liquid templates are usually a mix of at least two markup languages (Liquid markup itself plus your target format), they can get pretty ugly.

Reviewer: ```suggestion
Because Liquid templates are usually a mix of at least two markup languages (Liquid markup itself plus your target format), they can feel pretty busy.
```
Or something like this.... the word 'ugly' kind of feels harsh here.

Author: Heh, it's my own code, but point taken ;-).

Such is the nature of content templating.
Nevertheless, we can generate such elegant code this way.

Don't worry about the details of this code.
Basically what it does is iterate (“loop”) through the `posts.paid` array in the data file, creating an entry for each item it comes across (objects describing my past paid gigs).
As it churns, Liquid expresses the details of each object, mixing it in with some AsciiDoc markup you'll also recognize in the output just below.

.From _build/pages/resume.adoc -- AsciiDoc source file output after parsing
[source,asciidoc]
----
include::pages/resume.adoc[tags="example-snippet"]
----

> The above snippet includes just the AsciiDoc generated for the first two entries, though the `_build/pages/resume.adoc` file is a complete AsciiDoc document source.

Reviewer: ```suggestion
The snippet above includes just the AsciiDoc generated for the first two entries, though the `_build/pages/resume.adoc` file is a complete AsciiDoc document source.
```

[NOTE]

> Since this file is built and thus not tracked in the Git repository, to see the whole file you would have to clone and build the Codewriting repo, generating a copy into the `_build/pages/` directory.)

Reviewer: ```suggestion
Since this file is built and thus not tracked in the Git repository, to see the whole file you would have to clone and build the Codewriting repo, generating a copy into the `_build/pages/` directory.
```

[TIP]
Because that final, prebuilt source is generated before this blog, the above snippet is directly sourced from the resulting file, accessed while this blog post was being rendered and ensuring it always accurately reflects the _current_ state of the underlying data.
(See the <<sniptip,bonus explainer>> at the end of this post.)

Finally, I render that AsciiDoc file into rich-text output.

> The PDF document is generated by the killer link:http://prawnpdf.org/api-docs/2.0/[*Prawn* PDF utility], after which the link:https://jekyllrb.com/[Jekyll static site generator] spits it out as HTML.

Reviewer: As a reader, I am wondering why PDF comes before HTML. You might consider briefly addressing the 'why' of this here.

Both of these processes invoke the link:https://github.com/asciidoctor/asciidoctor[*Asciidoctor*] rendering engine, as detailed below in <<the-build>>.

== The Output

> Here are images from the relevant details of both rendered artifacts.

Reviewer: ```suggestion
Here are images of both rendered artifacts:
```

.HTML version
====
image::assets/images/screenshot-detail_resume-employment-html.png[]
====

.PDF version
====
image::assets/images/screenshot-detail_resume-employment-pdf.png[]
====

[[the-build]]
== The Build

Now that we see what happened, let's explore how it was _scripted_.

All of this activity is coordinated by link:https://www.ajyl.org/liquidoc[*LiquiDoc*], my free, open-source utility for combining AsciiDoc, YAML, and Liquid with Jekyll and other link:https://jamstack.org/[JAMstack services and utilities].

> LiquiDoc is a nascent API for the AJYL docstack, designed for building rich-text documents in many editions and formats, all from the same source, by invoking upstream tools (such as Asciidoctor, Liquid, and Jekyll processing engines).

Reviewer: ```suggestion
LiquiDoc is a nascent API for the AJYL docstack.
It is designed for building rich-text documents in many editions and formats, all from the same source, by invoking upstream tools (such as Asciidoctor, Liquid, and Jekyll processing engines).
```

This is a fab description, and it gets bogged down or convoluted by the long sentence.

The Codewriting project's LiquiDoc link:https://github.com/briandominick/codewriting/blob/master/_configs/build-global.yml[build configuration file (config)] for this site is pretty long, but essentially three relevant stages generate both the HTML and PDF versions of the resumé.

> Let's look at the first such stage.

Reviewer: ```suggestion
Let's look at the first stage.
```

\endif::backend-pdf[]
----

The first and last lines are conditionals; they ensure the enveloped settings are only applied during rendering procedures if the `backend-pdf` setting (“attribute” in AsciiDoc parlance) is defined, but they will replace any previously set parameters with the same names.

Which gets us back to the LiquiDoc config.
The next stage instructs LiquiDoc to actually render the PDF version of the output from the source file prebuilt in the last stage (`_build/pages/resume.adoc`).

.From _configs/build-global.yml -- LiquiDoc config file resume PDF render stage
[source,yaml]
----
include::assets/includes/build-global.yml[tags="render-resume-example-snippet"]
----

This block invokes a very simple Asciidoctor conversion using the link:https://github.com/asciidoctor/asciidoctor-pdf[PDF backend extension], which engages Prawn for the final conversion.
A customized link:https://github.com/briandominick/codewriting/blob/master/theme/pdf/resume-theme.yml[YAML-formatted theme file] enforces the style of the PDF output.

The resulting PDF artifact will later be copied into the Jekyll-generated website, but first we have to perform that Jekyll generation.

> The LiquiDoc config has us covered here, as well.

Reviewer: This sounds interesting but as I reader I'm not quite sure what you mean.

Author: Rewriting this a bit.

[[example-config-jekyll]]
.From _configs/build-global.yml -- Jekyll build execution stage
[source,yaml]
----
include::assets/includes/build-global.yml[tags="build-jekyll-example-snippet"]
----

This section of the config script instructs Jekyll to generate a site, one page of which will be the HTML version of the resumé, as dictated by the `_build/pages/resume.adoc` file's frontmatter settings referenced above.
This conversion does not require a specific instruction, as Jekyll renders all files in `_build/pages/` by default, and this prebuilt file is after all just another Jekyll-ready AsciiDoc file like link:https://github.com/briandominick/codewriting/tree/master/pages[all its siblings].

== Conclusion

> These are all the main features of an AJYL platform: small data in YAML flat files, content in AsciiDoc, templating and transformation with Liquid, and Jekyll to put it all together as a proper website.

Reviewer: LOVE THIS

The rest of the Codewriting site uses similar techniques to massage data and content together, either during prebuilding with Liquid or when rendering final artifacts.
Asciidoctor uses parameters (AsciiDoc attributes) when converting AsciiDoc files to HTML, PDF, and other formats.
Also during the render build, Jekyll engages Liquid's template engine to construct the site's HTML (or even to link:https://github.com/asciidoctor/jekyll-asciidoc#enabling-liquid-preprocessing[preprocess AsciiDoc content files containing Liquid markup]).

> All this complementarity is key to the AJYL environment.

Reviewer: Hmm, is `complementarity ` really a word?

Author: Yes.

> While a resumé is merely a novel implementation of this varied and powerful technology stack, imagine what this kind of flexibility can do for conveying complex product data in numerous permutations of editions and output formats.

Reviewer: Move this to the top. Rad.

Author: I don't want to leave the blog without a good conclusion sentence, and I don't have it in me to try to come up with another haha. But thanks!
